File: android-native/app/src/main/java/com/adamgoodwin/highlow/game/GameEngine.kt
package com.adamgoodwin.highlow.game

import kotlin.math.floor
import kotlin.math.min
import kotlin.random.Random

object GameEngine {
    private val suits = listOf("♠", "♥", "♦", "♣")
    val bonusConfig = BonusConfig()
    const val minBet = 10
    const val betStep = 10
    val quickBets = listOf(10, 50, 100, 500)
    const val shuffleThreshold = 10

    fun createDeck(deckCount: Int = 1): List<Card> =
        buildList {
            repeat(deckCount) { d ->
                suits.forEach { suit ->
                    (1..13).forEach { rank ->
                        add(Card(id = "$d-$suit-$rank-${size}", suit = suit, rank = rank))
                    }
                }
            }
        }

    fun shuffleDeck(deck: List<Card>, random: Random = Random.Default): List<Card> =
        deck.shuffled(random)

    private fun randomItem(cards: List<Card>, random: Random): Card = cards[random.nextInt(cards.size)]

    fun removeCard(deck: List<Card>, cardId: String): List<Card> = deck.filterNot { it.id == cardId }

    fun compareCards(current: Card, next: Card): Int = when {
        next.rank > current.rank -> 1
        next.rank < current.rank -> -1
        else -> 0
    }

    fun determineOutcome(choice: PlayerChoice, current: Card, next: Card): RoundOutcome {
        val cmp = compareCards(current, next)
        if (cmp == 0) return RoundOutcome.PUSH // Explicit tie behavior for MVP
        return when (choice) {
            PlayerChoice.HIGH -> if (cmp > 0) RoundOutcome.WIN else RoundOutcome.LOSS
            PlayerChoice.LOW -> if (cmp < 0) RoundOutcome.WIN else RoundOutcome.LOSS
        }
    }

    data class PayoutResult(val streak: Int, val bonus: Int, val profit: Int)

    fun resolvePayout(bet: Int, outcome: RoundOutcome, previousStreak: Int, config: BonusConfig = bonusConfig): PayoutResult {
        return when (outcome) {
            RoundOutcome.PUSH -> PayoutResult(previousStreak, 0, 0)
            RoundOutcome.LOSS -> PayoutResult(0, 0, -bet)
            RoundOutcome.WIN -> {
                val streak = previousStreak + 1
                val bonus = if (streak % config.streakEvery == 0) {
                    min(floor(bet * config.bonusPct).toInt(), config.bonusCap)
                } else 0
                PayoutResult(streak, bonus, bet + bonus)
            }
        }
    }

    data class DrawCurrentResult(val current: Card, val deck: List<Card>)

    fun drawCurrentCard(deck: List<Card>, mode: GameMode, random: Random = Random.Default): DrawCurrentResult {
        // Rigged modes avoid A/K as current card so HIGH/LOW can both be forceable.
        val eligible = if (mode == GameMode.FAIR) deck else deck.filter { it.rank in 2..12 }
        val source = if (eligible.isNotEmpty()) eligible else deck
        val card = randomItem(source, random)
        return DrawCurrentResult(card, removeCard(deck, card.id))
    }

    data class PickNextResult(
        val next: Card,
        val deck: List<Card>,
        val didReshuffle: Boolean,
        val riggedFallbackUsed: Boolean
    )

    fun pickNextCard(
        deck: List<Card>,
        current: Card,
        mode: GameMode,
        choice: PlayerChoice,
        deckCount: Int = 1,
        random: Random = Random.Default
    ): PickNextResult {
        if (mode == GameMode.FAIR) {
            val next = randomItem(deck, random)
            return PickNextResult(next, removeCard(deck, next.id), didReshuffle = false, riggedFallbackUsed = false)
        }

        val wantsHigher = (mode == GameMode.ALWAYS_WIN && choice == PlayerChoice.HIGH) ||
            (mode == GameMode.ALWAYS_LOSE && choice == PlayerChoice.LOW)
        val candidates = deck.filter { if (wantsHigher) it.rank > current.rank else it.rank < current.rank }
        if (candidates.isNotEmpty()) {
            val next = randomItem(candidates, random)
            return PickNextResult(next, removeCard(deck, next.id), didReshuffle = false, riggedFallbackUsed = false)
        }

        // Explicit fallback path for impossible deck states in rigged modes: reshuffle and try to force again.
        val reshuffled = shuffleDeck(removeCard(createDeck(deckCount), current.id), random)
        val forced = reshuffled.filter { if (wantsHigher) it.rank > current.rank else it.rank < current.rank }
        if (forced.isNotEmpty()) {
            val next = randomItem(forced, random)
            return PickNextResult(next, removeCard(reshuffled, next.id), didReshuffle = true, riggedFallbackUsed = false)
        }

        // Final fallback (should be unreachable with non-edge current-card guard). Avoid ties if possible.
        val nonTie = reshuffled.filter { it.rank != current.rank }
        val next = randomItem(if (nonTie.isNotEmpty()) nonTie else reshuffled, random)
        return PickNextResult(next, removeCard(reshuffled, next.id), didReshuffle = true, riggedFallbackUsed = true)
    }

    fun modeLabel(mode: GameMode): String = when (mode) {
        GameMode.FAIR -> "Fair"
        GameMode.ALWAYS_WIN -> "Demo: Always Win"
        GameMode.ALWAYS_LOSE -> "Chaos: Always Lose"
    }

    fun rankLabel(rank: Int): String = when (rank) {
        1 -> "A"
        11 -> "J"
        12 -> "Q"
        13 -> "K"
        else -> rank.toString()
    }
}

File: lib/game-engine.ts
import { BONUS_CONFIG, DEFAULT_TIE_BEHAVIOR } from "@/lib/constants";
import type { BonusConfig, Card, GameMode, PlayerChoice, RoundOutcome, Suit, TieBehavior } from "@/lib/types";

const SUITS: Suit[] = ["♠", "♥", "♦", "♣"];

export function createDeck(): Card[] {
  const deck: Card[] = [];
  for (const suit of SUITS) {
    for (let rank = 1; rank <= 13; rank += 1) {
      deck.push({ id: `${suit}-${rank}`, suit, rank });
    }
  }
  return deck;
}

export function shuffleDeck(deck: Card[], rng: () => number = Math.random): Card[] {
  const copy = [...deck];
  for (let i = copy.length - 1; i > 0; i -= 1) {
    const j = Math.floor(rng() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

function pickRandom<T>(items: T[], rng: () => number = Math.random): T {
  return items[Math.floor(rng() * items.length)];
}

export function removeCard(deck: Card[], cardId: string): Card[] {
  const idx = deck.findIndex((c) => c.id === cardId);
  if (idx < 0) return deck;
  return [...deck.slice(0, idx), ...deck.slice(idx + 1)];
}

export function compareCards(current: Card, next: Card): -1 | 0 | 1 {
  if (next.rank > current.rank) return 1;
  if (next.rank < current.rank) return -1;
  return 0;
}

export function determineOutcome(
  choice: PlayerChoice,
  current: Card,
  next: Card,
  tieBehavior: TieBehavior = DEFAULT_TIE_BEHAVIOR
): RoundOutcome {
  const cmp = compareCards(current, next);
  if (cmp === 0) return tieBehavior === "push" ? "push" : "loss";
  if (choice === "high") return cmp > 0 ? "win" : "loss";
  return cmp < 0 ? "win" : "loss";
}

export function calculateBonus(bet: number, streakAfterRound: number, config: BonusConfig = BONUS_CONFIG) {
  if (streakAfterRound <= 0) return 0;
  if (streakAfterRound % config.streakEvery !== 0) return 0;
  return Math.min(Math.floor(bet * config.bonusPct), config.bonusCap);
}

export function resolvePayout(params: { bet: number; outcome: RoundOutcome; previousStreak: number; bonusConfig?: BonusConfig }) {
  const { bet, outcome, previousStreak, bonusConfig = BONUS_CONFIG } = params;
  if (outcome === "push") return { streak: previousStreak, bonus: 0, profit: 0 };
  if (outcome === "loss") return { streak: 0, bonus: 0, profit: -bet };
  const streak = previousStreak + 1;
  const bonus = calculateBonus(bet, streak, bonusConfig);
  return { streak, bonus, profit: bet + bonus };
}

export function drawCurrentCard(params: { deck: Card[]; mode: GameMode; rng?: () => number }) {
  const { deck, mode, rng = Math.random } = params;
  // Rigged modes avoid edge ranks (A/K) so both HIGH and LOW can always be forced next.
  const eligible = mode === "fair" ? deck : deck.filter((c) => c.rank > 1 && c.rank < 13);
  const source = eligible.length ? eligible : deck;
  const card = pickRandom(source, rng);
  return { card, deck: removeCard(deck, card.id) };
}

export interface PickNextCardResult {
  nextCard: Card;
  deck: Card[];
  didReshuffle: boolean;
  riggedFallbackUsed: boolean;
}

export function pickNextCard(params: {
  deck: Card[];
  current: Card;
  mode: GameMode;
  choice: PlayerChoice;
  rng?: () => number;
}): PickNextCardResult {
  const { deck, current, mode, choice, rng = Math.random } = params;

  if (mode === "fair") {
    const nextCard = pickRandom(deck, rng);
    return { nextCard, deck: removeCard(deck, nextCard.id), didReshuffle: false, riggedFallbackUsed: false };
  }

  const wantsHigher = (mode === "alwaysWin" && choice === "high") || (mode === "alwaysLose" && choice === "low");
  const candidates = deck.filter((c) => (wantsHigher ? c.rank > current.rank : c.rank < current.rank));
  if (candidates.length) {
    const nextCard = pickRandom(candidates, rng);
    return { nextCard, deck: removeCard(deck, nextCard.id), didReshuffle: false, riggedFallbackUsed: false };
  }

  // Fallback path is explicit and deterministic in code:
  // reshuffle a fresh deck (minus current card) and try to force the required relation again.
  const reshuffled = shuffleDeck(removeCard(createDeck(), current.id), rng);
  const forcedCandidates = reshuffled.filter((c) => (wantsHigher ? c.rank > current.rank : c.rank < current.rank));
  if (forcedCandidates.length) {
    const nextCard = pickRandom(forcedCandidates, rng);
    return { nextCard, deck: removeCard(reshuffled, nextCard.id), didReshuffle: true, riggedFallbackUsed: false };
  }

  // Last resort (should be unreachable due to drawCurrentCard edge-rank guard).
  const nonTie = reshuffled.filter((c) => c.rank !== current.rank);
  const nextCard = pickRandom(nonTie.length ? nonTie : reshuffled, rng);
  return { nextCard, deck: removeCard(reshuffled, nextCard.id), didReshuffle: true, riggedFallbackUsed: true };
}

export function modeLabel(mode: GameMode) {
  if (mode === "fair") return "Fair";
  if (mode === "alwaysWin") return "Demo: Always Win";
  return "Chaos: Always Lose";
}
